ballerina/main.bal

```diff
@@ -124,25 +121,83 @@ isolated function parseResponseAsJson(string resp) returns json|error {
     return result;
 }
 
-isolated function parseResponseAsType(string resp, typedesc<anydata> expectedResponseTypedesc) returns anydata|error {
-    json respJson = check parseResponseAsJson(resp);
-    anydata|error result = trap respJson.fromJsonWithType(expectedResponseTypedesc);
+isolated function parseResponseAsType(string resp,
+        typedesc<anydata> expectedResponseTypedesc, boolean isOriginallyJsonObject) returns anydata|error {
+    if !isOriginallyJsonObject {
+        map<json> respContent = check resp.fromJsonStringWithType();
+        anydata|error result = trap respContent[RESULT].fromJsonWithType(expectedResponseTypedesc);
+        if result is error {
+            return handleParseResponseError(result);
+        }
+        return result;
+    }
+
+    anydata|error result = check resp.fromJsonStringWithType(expectedResponseTypedesc);
     if result is error {
         return handleParseResponseError(result);
     }
     return result;
 }
 
 isolated function handleParseResponseError(error chatResponseError) returns error {
-    if chatResponseError.message().includes(JSON_CONVERSION_ERROR) 
+    if chatResponseError.message().includes(JSON_CONVERSION_ERROR)
             || chatResponseError.message().includes(CONVERSION_ERROR) {
         return error(string `${ERROR_MESSAGE}`, detail = chatResponseError);
     }
     return chatResponseError;
 }
 
-isolated function getExpectedResponseSchema(typedesc<anydata> expectedResponseTypedesc) returns map<json> {
+isolated function getExpectedResponseSchema(typedesc<anydata> expectedResponseTypedesc) returns SchemaResponse {
     // Restricted at compile-time for now.
     typedesc<json> td = checkpanic expectedResponseTypedesc.ensureType();
-    return generateJsonSchemaForTypedescAsJson(td);
+    return generateJsonObjectSchema(generateJsonSchemaForTypedescAsJson(td));
+}
+
+isolated function generateJsonObjectSchema(map<json> schema) returns SchemaResponse {
+    string[] supportedMetaDataFields = ["$schema", "$id", "$anchor", "$comment", "title", "description"];
+
+    if schema.hasKey("type") {
+        if schema["type"] == "object" {
+            return {schema};
+        }
+    }
+
+    map<json> updatedSchema = map from var [key, value] in schema.entries()
+        where supportedMetaDataFields.indexOf(key) is int
+        select [key, value];
+
+    updatedSchema["type"] = "object";
+    map<json> content = map from var [key, value] in schema.entries()
+        where supportedMetaDataFields.indexOf(key) !is int
+        select [key, value];
+
+    updatedSchema["properties"] = {[RESULT]: content};
+
+    return {schema: updatedSchema, isOriginallyJsonObject: false};
+}
+
+isolated function getToolCallingDescription() returns string {
```

Comment:
Can be an expression-bodied function. Please make sure we're writing the ideal Ballerina code.